pkg/registry/core/service/storage/storage_test.go

```diff
@@ -5832,4 +5834,979 @@ func TestDeleteDryRun(t *testing.T) {
 	}
 }
 
-//FIXME: Tests for update()
+type svcTestCase struct {
+	svc         *api.Service
+	expectError bool
+
+	// We could divine these by looking at the Service, but that's a
+	// vector for test bugs and more importantly it makes the test cases less
+	// self-documenting.
+	expectClusterIPs          bool
+	expectHeadless            bool
+	expectNodePorts           bool
+	expectHealthCheckNodePort bool
+}
+
+func callName(before, after *api.Service) string {
+	if before == nil {
+		return "create"
+	}
+	return "update"
+}
+
+func proveClusterIPsAllocated(t *testing.T, storage *GenericREST, before, after *api.Service) {
+	t.Helper()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		if sing, plur := after.Spec.ClusterIP, after.Spec.ClusterIPs[0]; sing != plur {
+			t.Errorf("%s: expected clusterIP == clusterIPs[0]: %q != %q", callName(before, after), sing, plur)
+		}
+	}
+
+	clips := []string{}
+	if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+		clips = after.Spec.ClusterIPs
+	} else {
+		clips = append(clips, after.Spec.ClusterIP)
+	}
+	for _, clip := range clips {
+		if !ipIsAllocated(t, storage.alloc.serviceIPAllocatorsByFamily[familyOf(clip)], clip) {
+			t.Errorf("%s: expected clusterIP to be allocated: %q", callName(before, after), clip)
+		}
+	}
+
+	for i, fam := range after.Spec.IPFamilies {
+		if want, got := fam, familyOf(after.Spec.ClusterIPs[i]); want != got {
+			t.Errorf("clusterIP is the wrong IP family: want %s, got %s", want, got)
+		}
+	}
+
+	if before != nil {
+		if before.Spec.ClusterIP != "" {
+			if want, got := before.Spec.ClusterIP, after.Spec.ClusterIP; want != got {
+				t.Errorf("wrong clusterIP: wanted %q, got %q", want, got)
+			}
+		}
+		for i := range before.Spec.ClusterIPs {
+			if want, got := before.Spec.ClusterIPs[i], after.Spec.ClusterIPs[i]; want != got {
+				t.Errorf("wrong clusterIPs[%d]: wanted %q, got %q", i, want, got)
+			}
+		}
+		for i := range before.Spec.IPFamilies {
+			if want, got := before.Spec.IPFamilies[i], after.Spec.IPFamilies[i]; want != got {
+				t.Errorf("wrong IPFamilies[%d]: wanted %q, got %q", i, want, got)
+			}
+		}
+	}
+}
+
+func proveClusterIPsDeallocated(t *testing.T, storage *GenericREST, before, after *api.Service) {
+	t.Helper()
+
+	if after != nil && after.Spec.ClusterIP != api.ClusterIPNone {
```

Comment:
bailout early. by having
```
if after != nil && after.Spec.ClusterIP == api.ClusterIPNone {
return
}
```
instead of the double ifs guarding both assertion.

Reply:
but I don't want to return - I want to check before, too.  Maybe I miss your point?

Reply:
We can ignore this for now.